src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioActiveDocumentTracker.cs

```diff
@@ -0,0 +1,290 @@
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public int OnElementValueChanged([ComAliasName("Microsoft.VisualStudio.Shell.Interop.VSSELELEMID")]uint elementid, object varValueOld, object varValueNew)
+        {
+            AssertIsForeground();
+
+            if (elementid == (uint)VSConstants.VSSELELEMID.SEID_DocumentFrame)
+            {
+                // Remember the newly activated frame so it can be read from another thread.
+                _activeFrame = varValueNew as IVsWindowFrame;
+
+                if (_activeFrame != null)
+                {
+                    // This might be the first time we've ever seen this IVsWindowFrame, so add it to our tracking too
+                    TrackWindowFrame(_activeFrame);
+                }
+
+                this.DocumentsChanged?.Invoke(this, EventArgs.Empty);
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        public int OnCmdUIContextChanged([ComAliasName("Microsoft.VisualStudio.Shell.Interop.VSCOOKIE")]uint dwCmdUICookie, [ComAliasName("Microsoft.VisualStudio.OLE.Interop.BOOL")]int fActive)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public event EventHandler<EventArgs> NonRoslynBufferTextChanged;
+
+        /// <summary>
+        /// Listens to frame notifications for a visible frame. When the frame becomes invisible or closes,
+        /// then it automatically disconnects.
+        /// </summary>
+        private class FrameListener : IVsWindowFrameNotify, IVsWindowFrameNotify2
+        {
+            public readonly IVsWindowFrame Frame;
+
+            private readonly VisualStudioActiveDocumentTracker _documentTracker;
+            private readonly uint _frameEventsCookie;
+
+            private readonly ITextBuffer _textBuffer;
+
+            public FrameListener(VisualStudioActiveDocumentTracker service, IVsWindowFrame frame)
+            {
+                _documentTracker = service;
+                _documentTracker.AssertIsForeground();
+
+                this.Frame = frame;
+
+                ((IVsWindowFrame2)frame).Advise(this, out _frameEventsCookie);
+
+                if (ErrorHandler.Succeeded(frame.GetProperty((int)__VSFPROPID.VSFPROPID_DocData, out var docData)))
+                {
+                    if (docData is IVsTextBuffer bufferAdapter)
+                    {
+                        _textBuffer = _documentTracker._editorAdaptersFactoryService.GetDocumentBuffer(bufferAdapter);
+
+                        if (!_textBuffer.ContentType.IsOfType(ContentTypeNames.RoslynContentType))
+                        {
+                            _textBuffer.Changed += NonRoslynTextBuffer_Changed;
+                        }
+                    }
+                }
+            }
+
+            private void NonRoslynTextBuffer_Changed(object sender, TextContentChangedEventArgs e)
+            {
+                _documentTracker.DocumentsChanged?.Invoke(_documentTracker, EventArgs.Empty);
```

Comment:
interesting.  can you explain why we want this?  why do we want ot hear about chnages to non-roslyn docs?

Reply:
(Explained at the other place you asked the question.)